Src/GoogleApis.Auth.DotNet4/OAuth2/DefaultCredentialProvider.cs

```diff
@@ -0,0 +1,252 @@
+
+        /// <summary>
+        /// Returns the Application Default Credentials
+        /// See <see cref="M:Google.Apis.Auth.OAuth2.GoogleAuth.GetApplicationDefaultCredential"/> for details.
+        /// </summary>
+        public GoogleCredential GetApplicationDefaultCredential()
+        {
+            if (cachedCredential == null)
+            {
+                lock (lockObject)
+                {
+                    if (cachedCredential == null)
+                    {
+                        cachedCredential = GetDefaultCredentialUnsynchronized();
+                    }
+                }
+            }
+            if (cachedCredential != null)
+            {
+                return cachedCredential;
+            }
+
+            throw new InvalidOperationException(String.Format("The Application Default Credentials are not available. They are available if running"
+                + " in Google Compute Engine. Otherwise, the environment variable {0} must be defined"
+                + " pointing to a file defining the credentials. See {1} for more information.",
+                CredentialEnvironmentVariable,
+                HelpPermalink));
+        }
+
+        /// <summary>Unsynchronized helper method to obtain the Application Default Credentials.</summary>
+        private GoogleCredential GetDefaultCredentialUnsynchronized()
+        {
+            string credentialPath = null;
+            
+            // 1. First try the environment variable.
+            try
+            {
+                credentialPath = GetEnvironmentVariable(CredentialEnvironmentVariable);
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+            }
+            catch (Exception e)
+            {
+                // Catching generic exception type because any corrupted file could manifest in different ways including 
+                // but not limited to the System, System.IO or from the Newtonsoft.Json namespace.
+
+                throw new InvalidOperationException(String.Format("Error reading credential file from location {0}: {1}"
+                                        + "\nPlease check the value of the Environment Variable {2}",
+                                        credentialPath,
+                                        e.Message,
+                                        CredentialEnvironmentVariable));
+            }
+
+            // 2. Then try the well known file.
+            try
+            {
+                credentialPath = GetWellKnownCredentialFilePath();
+                if (!String.IsNullOrWhiteSpace(credentialPath))
+                {
+                    return CreateDefaultCredentialFromFile(credentialPath);
+                }
+            }
+            catch (FileNotFoundException)
+            {
+                // File is not present, eat the exception and move on to the next check.
```

Comment:
I think this should remain silent as when running on GCE you would always see a warning that the well known file hasn't been found  - but on GCE it's not expected to be there, so this would be misleading.